### DISCUSSION TITLE
Is there a way to trim all columns at once with ORM Query?

### DISCUSSION THREAD

**Discussion by miyamoto-yosuke** [sqlalchemy/sqlalchemy#7776] [Usage Questions]
When using orm, in the case of char type column, it is padded with whitespace after the column according to the item length.

```
      ・
      ・
      ・
   return db.query(Table).all()
      ・
      ・
      ・
```

```
{
  "data": {
    "char_column1": "char1                          ",
    "char_column2": "char2                          ",
    "char_column3": "char3                          ",
    "char_column4": "char4                          ",
    "char_column5": "char5                          ",
         ・
         ・
         ・
  }
}
```

I Hope.
```
{
  "data": {
    "char_column1": "char1",
    "char_column2": "char2",
    "char_column3": "char3",
    "char_column4": "char4",
    "char_column5": "char5",
         ・
         ・
         ・
  }
}
```

Is there a way to trim the whitespace behind?

**Comment by zzzeek** [comment]
sure, use [TypeDecorator](https://docs.sqlalchemy.org/en/14/core/custom_types.html#typedecorator-recipes)

```py
from sqlalchemy import TypeDecorator, CHAR

class TrimChar(TypeDecorator):
    impl = CHAR
    cache_key = True

    def process_result_value(self, value, dialect):
        if value:
            value = value.strip()
        return value
```

**Reply by CaselIT** [reply]
also if it's a new system it's probably worth considering using varchar instead of the fixed length versions

**Reply by miyamoto-yosuke** [reply]
@CaselIT
thank you. I plan to use postgresql on my new system.

@zzzeek
thank you.
Does TypeDecorator (TrimChar) have write code for each column? If the code is written more, I will try to handle it with frontend.

**Reply by CaselIT** [reply]
> @CaselIT thank you. I plan to use postgresql on my new system.

I would suggest using Text then on postgresql

**Reply by zzzeek** [reply]
> 
> @zzzeek thank you. Does TypeDecorator (TrimChar) have write code for each column? If the code is written more, I will try to handle it with frontend.

yes wherever you have Column(CHAR), you'd change it to Column(TrimChar)

**Reply by miyamoto-yosuke** [reply]
Thanks!

**Reply by grofte** [reply]
So if I understand this correctly you would have the cache_key disabled in most cases, only enabling it if you would have the same input values for trim many times, right? And you would turn it off for most applications? Or is this an entirely different kind of cache? Default it off but SQLAlchemy prompts you to set it explicitly when you use this class. 

```python
from sqlalchemy import TypeDecorator, CHAR

class TrimChar(TypeDecorator):
    impl = CHAR
    cache_key = False

    def process_result_value(self, value, dialect):
        if value:
            value = value.strip()
        return value
```

**Reply by CaselIT** [reply]
No, you would set this to True, since that cache refer to the sql string being rendered. Since `TrimChar` does not do any change to the sql emitted, you can safely set it to True.

You would set it to False if, for example, you had a `TypeDecorator` that depending on a parameter would make a column render sql as either `col_name` or `lower(col_name)`. In this case an sql statement that used such type decorator could not be cached

**Reply by zzzeek** [reply]
i've edited my answer to include cache_key=True
